### DISCUSSION TITLE
Any good libraries to get tasks from taskwarrior 3.0?

### DISCUSSION THREAD

**Discussion by djotaku** [GothenburgBitFactory/taskwarrior#3703] [Q&A]
Here's why I'm asking on here - while I appreciate how the dev team has worked on updating the tool database, too many (most?) tools and APIs that claim to support tw3 do not. They just didn't update their repo to say taskwarrior2. 

I would prefer Python or Golang (interestingly, almost every tool written for go in the tool database seems to be long abandoned), but could probably work with Rust if needed. 

All I need is a way to get the data in and out of the new format so that I can update my taskwarrior_web software to work with tw3. It seems that most likely the best way to do the sync portion is not via taskchampion (at least not yet), but with a program that functions equivalent to Dropbox.

Thanks!

**Comment by felixschurk** [comment]
Hei,
I think https://github.com/GothenburgBitFactory/tasklib should do the job as it interacts over the json import and export I assume.

At least taskwiki is also working with task v3 and to my understanding it uses tasklib in the backend.

**Reply by djotaku** [reply]
interestingly their repo is marked as taskwarrior2

**Reply by djmitche** [reply]
That might have just been a "safe" assumption without testing against TW3.

**Comment by djmitche** [comment]
It looks like that's https://github.com/djotaku/taskwarrior_web? I'm excited to have a web UI for Taskwarrior! Enabling this was one of the reasons to build things on Taskchampion.

[Taskchampion](https://github.com/GothenburgBitFactory/taskchampion) is the implementation of the data storage (and sync) support, so that's a good option. It's in Rust! There's also [semi-completed work to build a Python interface](https://github.com/GothenburgBitFactory/taskchampion/issues/385). Maybe you could finish that up, either as https://github.com/GothenburgBitFactory/taskchampion-py/ or just a part of your own repo, wrapping just the bits you need.

I'm unclear why you wouldn't want to do sync with Taskchampion. Syncing a replica is as simple as calling `replica.sync()?`! Trying to synchronize databases using things like Dropbox is going to lead to unhappiness, as such tools can't handle conflicting modifications. We just saw a case of that in #3695 (albiet that was using Syncthing to sync the taskchampion-sync-server DB, but the effect is the same). 

I think the rough outline for how a web UI would work is:
 * Keep a local Replica for each user. Currently that would need to be in SQLite, but adding a MySQL or Postgres backend shouldn't be hard.
 * Somehow get each user's sync config/credentials. If this is just a one-person app, then maybe that's best just put in the configuration. If you're supporting multiple people, perhaps those credentials can be in the app's database. Or, you could integrate a sync server into the app itself and use the app's credentials. You could also run the `taskchampion-sync-server` binary in the same podman container as your app.
 * Figure out when to sync -- I see that the existing repo does this with a crontask, and that makes sense for one user or a few users. If the app supports 1000's of users, maybe not!
